testsuite: remove sched-dummy, migrate testing to sched-simple

chu11 opened the pull request:
per discussion in #3455 

- some misc fixups, including a pretty big bug i found in some tests
- made `single` and `unlimited` mode tests more consistent to each other in the testsuite.  Unsure of legacy reasons why tests were different (e.g. in `single` mode `resource_summary` annotation was not output in `sched-dummy`).
- add a LOG_DEBUG message to `hello_cb` in `sched-simple`.  Could be made to output only on debugging circumstances, but thought it was ok to leave in for all circumstances.
- add a DEBUG_FAIL_ALLOC test hook to `sched-simple` to automatically fail the next allocation
- add a DEBUG_ANNOTATE_REASON_PENDING hook to inform `sched-simple` to annotate with `reason_pending` and `jobs_ahead` annotations
- update all tests to use `sched-simple` over `sched-dummy`
- remove `sched-dummy` and rename some files

grondo commented:
Before I delve in (later), noticed inconsistent use of capitalization in commit messages, e.g. `￼sched-simple: Fix code tabbing`, vs `￼sched-simple: fix error message typo`. (I'm sure I'm guilty of this as well). I prefer the 2nd (no capitalization), since commit subject is known not to be a full sentence anyway.

chu11 commented:
rebased and re-pushed.  I had a number of builder failures with loading/unloading modules, which I couldn't figure out.  Possibly something racy I was missing.

Then it occurred to me, with the removal `sched-dummy`, I no longer have to manually load/unload `job-ingest`, `job-info`, `job-manager`, I can run under the `job` test personality instead.  So I adjusted.  We'll see if builders work this time.

chu11 commented:
one builder failed w/ #3463 and another failed with a seemingly random collection of `flux-mini` failures.  I'd be worried it was something I did if it weren't for #3463 coming up at same time today.  In the `centos8 - py3.7` builder

```
Run src/test/checks-annotate.sh
Warning: Found 3 errors from 259 tests in testsuite
Error: not ok 8 - flux-mini batch: submit a series of jobs
Error: not ok 9 - flux-mini batch: job results are expected
Error: ERROR: t2701-mini-batch.t - exited with status 1
./t/t2701-mini-batch.output
./t/t2701-mini-batch.log
Error: not ok 30 - flux-shell: shell can launch flux
Error: ERROR: t2601-job-shell-standalone.t - exited with status 1
./t/t2601-job-shell-standalone.log
Error: not ok 7 - flux-mini alloc runs one broker per node by default
Error: ERROR: t2702-mini-alloc.t - exited with status 1
./t/t2702-mini-alloc.output
./t/t2702-mini-alloc.log
```

```
  expecting success: 
  	id1=$(flux mini batch --flags=waitable -n1 batch-script.sh) &&
  	id2=$(flux mini batch --flags=waitable -n4 batch-script.sh) &&
  	id3=$(flux mini batch --flags=waitable -N2 -n4 batch-script.sh) &&
  	run_timeout 60 flux job wait --verbose --all
  
  f2k6U2Yo: job completed successfully
  f2cqRYWs: job completed successfully
  not ok 8 - flux-mini batch: submit a series of jobs
  
  expecting success: 
  	test_debug "grep . flux-*.out" &&
  	grep "size=1 nodes=1" flux-${id1}.out &&
  	grep "size=1 nodes=1" flux-${id2}.out &&
  	grep "size=2 nodes=2" flux-${id3}.out
  
  size=1 nodes=1
  size=1 nodes=1
  not ok 9 - flux-mini batch: job results are expected
```

(no .output file?)
```
  FAIL: t2601-job-shell-standalone.t 30 - flux-shell: shell can launch flux
  #	
  #		flux jobspec srun -N1 -n8 flux start flux getattr size >j8flux &&
  #		${FLUX_SHELL} -v -s -r 0 -j j8flux -R R8 39 \
  #			>flux.out 2>flux.err &&
  #		grep -x "0: 8" flux.out
  #	
```

```
expecting success: 
  	$runpty -o multi.out flux mini alloc -n5 flux lsattr -v &&
  	test_debug "cat multi.out" &&
  	grep "size  *2" multi.out
  
  runpty: INFO: child exited with status 0x100
  not ok 7 - flux-mini alloc runs one broker per node by default`

```

grondo commented:
I'm going to try restarting tests. It seemed like maybe Github runners were behaving strangely late last week.

Then it seems like it would be better to get this cleanup in sooner rather than later, so I'll take an immediate look.

grondo commented:
Hm, there seems to be something wrong with the scheduler on this branch. I have a little job throughput test which reproducibly gets ~70 job/s on current master, but on this branch I'm seeing about 6 jobs/s.

It could be something in my environment, though I'm running both tests from different build directories on fluke.

However, @chu11, maybe you can try this out and see if the problem reproduces for you?

```python
#!/usr/bin/env python3
#
#
import sys
import time
import argparse

import flux
from flux import job
from flux.job import JobspecV1, JobID

try:
    from flux.progress import Bottombar
except:
    pass

parser = argparse.ArgumentParser(description="Run job throughput test")
parser.add_argument(
    "-n",
    "--njobs",
    type=int,
    metavar="N",
    help="Set the total number of jobs to run",
    default=100,
)
parser.add_argument(
    "-x",
    "--exec",
    help="Do not simulate execution, actually run jobs",
    action="store_true",
)
parser.add_argument(
    "-s",
    "--status",
    help="Add a status bar to bottom of terminal",
    action="store_true",
)
parser.add_argument(
    "-v",
    "--verbose",
    help="Log job events",
    action="store_true",
)
parser.add_argument("command", nargs=argparse.REMAINDER, default=["true"])

args = parser.parse_args()


class BulkRun:
    def __init__(self, handle, total, jobspec):
        self.handle = handle
        self.total = total
        self.jobspec = jobspec
        self.jobs = {}
        self.submitted = 0
        self.running = 0
        self.complete = 0
        self.bb = None

    def statusline(self, bb, width):
        return (
            f"{self.total:>6} jobs: {self.submitted:>6} submitted, "
            f"{self.running:>6} running, {self.complete} completed"
        )

    def event_cb(self, future, args, jobid):
        event = future.get_event()
        if event is not None:
            if args.verbose:
                print(f"{jobid}: {event.name}")
            if event.name == "submit":
                self.submitted += 1
                self.jobs[jobid][event.name] = event
            elif event.name == "start":
                self.running += 1
            elif event.name == "finish":
                self.running -= 1
                self.complete += 1
            elif event.name == "clean":
                self.jobs[jobid][event.name] = event
            if self.bb:
                self.bb.update()

    def submit_cb(self, future, args):
        try:
            jobid = JobID(future.get_id())
            self.jobs[jobid] = {}
            fut = job.event_watch_async(self.handle, jobid)
            fut.then(self.event_cb, args, jobid)
        except Exception as exc:
            print(f"Submission failed: {exc}", file=sys.stderr)

    def run(self, args):
        if args.status:
            self.bb = Bottombar(self.statusline).start()
        spec = self.jobspec.dumps()
        for i in range(0, args.njobs):
            job.submit_async(self.handle, spec).then(self.submit_cb, args)
        self.handle.reactor_run()

        if self.bb:
            self.bb.stop()

        return self


t0 = time.time()

#  Create a test jobspec
if not args.command:
    args.command = ["true"]
jobspec = JobspecV1.from_command(args.command)
if not args.exec:
    jobspec.setattr("system.exec.test.run_duration", "0.001s")

bulk = BulkRun(flux.Flux(), args.njobs, jobspec).run(args)

jobs = bulk.jobs

#  Get the job with the earliest 'submit' event:
first = jobs[min(jobs.keys(), key=lambda x: jobs[x]["submit"].timestamp)]

#  Get the job with the latest 'clean' event:
last = jobs[max(jobs.keys(), key=lambda x: jobs[x]["clean"].timestamp)]

script_runtime = time.time() - t0
job_runtime = last["clean"].timestamp - first["submit"].timestamp
jps = args.njobs / job_runtime
jpsb = args.njobs / script_runtime

print(f"number of jobs: {args.njobs}")
print(f"script runtime: {script_runtime:<6.3f}s")
print(f"job runtime:    {job_runtime:<6.3f}s")
print(f"throughput:     {jps:<.1f} job/s (script: {jpsb:5.1f} job/s)")
```

On master:
```
$ src/cmd/flux start sh -c 'flux module reload sched-simple unlimited; flux python ../flux-core.git/throughput.py -n 1024'
number of jobs: 1024
script runtime: 14.855s
job runtime:    14.706s
throughput:     69.6 job/s (script:  68.9 job/s)
```

On this branch:
```
$  src/cmd/flux start sh -c 'flux module reload sched-simple unlimited; flux python ../flux-core.git/throughput.py -n 1024'
number of jobs: 1024
script runtime: 184.284s
job runtime:    154.650s
throughput:     6.6 job/s (script:   5.6 job/s)
```

It seems to get much worse the more jobs are enqueued, so something in this PR is not scaling at all with the number of submitted jobs in the "unlimited" case.

grondo commented on src/modules/sched-simple/sched.c:
```diff
@@ -21,6 +21,13 @@
+// e.g. flux module debug --clearbit 0x1 sched-simple
+enum module_debug_flags {
+    DEBUG_FAIL_ALLOC = 1, // while set, alloc requests fail
+    DEBUG_ANNOTATE_REASON_PENDING = 1, // add reason_pending annotation
```
Are both these enum values supposed to be the same (i.e. `1`)?

chu11 commented:
@grondo hmmm, my initial thought is I'm not checking the debug/test flags correctly.  Annotations will loop through the queue of jobs regularly, definitely slowing things down.  Which is why I (intended) for them to be run only during debug/test scenarios.

grondo commented on src/modules/sched-simple/sched.c:
```diff
@@ -208,6 +225,28 @@ static int try_alloc (flux_t *h, struct simple_sched *ss)
+{
+    int jobs_ahead = 0;
+
+    if (flux_module_debug_test (ss->h, DEBUG_ANNOTATE_REASON_PENDING, false))
```
Ah, this could be the problem with the impacted throughput, shouldn't this be `if (!flux_module_debug_test (...)`?

chu11 replied:
doh! yup, and in combo with the bitmask mistake above, that's why tests were passing.

grondo reviewed (changes requested):
There seems to be a couple problems with the new module debug flags. That might explain the performance impact of this PR (code is walking the scheduler queue on every iteration to calculate debugging data).

chu11 commented on src/modules/sched-simple/sched.c:
```diff
@@ -379,8 +423,17 @@ static int hello_cb (flux_t *h,
 
-    struct rlist *alloc = rlist_from_R (R);
+    flux_log (h, LOG_DEBUG,
+              "hello: id=%ju priority=%d userid=%u t_submit=%0.1f R=%s",
```
oh yup, I cut & paste that from the `sched-dummy` code.  Yeah, we can definitely drop `R`.

chu11 replied:
oh wait, i'm getting my PRs mixed up.  At this exact moment in time `priority` is typed as an `int` in `libschedutil`, so I think '%d' is correct for now.  This should get fixed up via #3450, which I'm currently working on and will attach to PR #3447.

chu11 commented:
rebased & re-pushed per comments above from @grondo.  The stupid mistakes in `sched-simple` were definitely the cause of the slowdown.

Due to the fixes, I did have to add a race fix in the tests.

```
        flux queue stop &&
        flux module load sched-simple &&
        flux module debug --setbit 0x2 sched-simple &&
        flux queue start
```

We need to make sure jobs aren't scheduled until the debug bits are set, so had to add a queue stop/start around it.

grondo commented:
Great! I can verify that fixed the issue.

grondo commented:
LGTM now @chu11. Go ahead and set MWP when ready. (I'll restart tests, looks like one timed out)

chu11 commented:
@grondo, thanks!
